File: utils/XeThru_utils/xeX4Thru_software/ModuleConnector/ModuleConnector-unix-1/matlab/examples/X4M200_X4M300_manipulate_noisemap.m
% This example illustrates how to manipulate stored noisemaps on the module.
%
% To complete the following example you need:
% - An X4M200 or X4M300 module
% - The ModuleConnector library
% - MATLAB


%add paths
addModuleConnectorPath();
% if running on a 32-bit Windows system, instead run:
% addModuleConnectorPath('win32');

clc
clear


% Specify parameters
COMPORT             = 'COM4';   % Device com-port
options.interface   = 'x4m300';         % Interface to use. x4m300 or x4m200

Lib = ModuleConnector.Library;

clear mc; mc = ModuleConnector.ModuleConnector(COMPORT, 0);
pause(1);

% Get the XEP interface
xep = mc.get_xep();

% Get app interface and set profile
if strcmp( options.interface, 'x4m200' )
    app = mc.get_x4m200();
    % Adult respiration (respiration2) profile.
    profile = hex2dec('064e57ad');
elseif strcmp( options.interface, 'x4m300' )
    app = mc.get_x4m300();
    % Adult presence (presence2) profile.
    profile = hex2dec('014d4ab8');
else
    error('Interface not recognized.\n')
end

% Load a profile. Noisemap controls are stored associated with this profile,
% meaning setting/getting it witout loading a profile doesn't work.
status = app.load_profile(profile);
if status, error('Failed loading profile.'); end

% Set a detection zone. The noisemap requires a matching detection zone.
status = app.set_detection_zone(1, 3);
if status, error('Failed setting detection zone.'); end

[ctrl, status] = app.get_noisemap_control();
if status, error('Failed getting noisemap control.'); end

% Enable use of noisemap
ctrl = bitor(ctrl, 1);
% Use adaptive
ctrl = bitor(ctrl, 2);
% Don't reinitialize noisemap if stored, and store it after initialization if
% not present.
ctrl = bitand(ctrl, bitcmp(uint32(4)));
status = app.set_noisemap_control(ctrl);
if status, error('Failed setting noisemap control.'); end

% Delete old noisemap if any, forcing it to initialize and store a new one.
files = xep.find_all_files();
if ismember(xep.XTFILE_TYPE_NOISEMAP_FAST, files)
    xep.delete_file(xep.XTFILE_TYPE_NOISEMAP_FAST, 0);
end
if ismember(xep.XTFILE_TYPE_NOISEMAP_SLOW, files)
    xep.delete_file(xep.XTFILE_TYPE_NOISEMAP_SLOW, 0);
end

disp('Initializing a noisemap. Vacate the premises.');
pause(5);
status = app.set_sensor_mode('run');
if status, error('Failed starting sensor.'); end
disp('waiting for 130s ...');
pause(130); % 10 seconds margin of error
status = app.set_sensor_mode('stop');
if status, error('Failed stopping sensor.'); end
disp('Done.')
pause(1);

files = xep.find_all_files();
if ~ismember(xep.XTFILE_TYPE_NOISEMAP_FAST, files)
    error('Did not store a fast noisemap');
end
if ~ismember(xep.XTFILE_TYPE_NOISEMAP_SLOW, files)
    error('Did not store a slow noisemap');
end

nm_fast = xep.get_file(xep.XTFILE_TYPE_NOISEMAP_FAST, 0);
nm_slow = xep.get_file(xep.XTFILE_TYPE_NOISEMAP_SLOW, 0);

% Do it once more, but with a smaller detection zone.

status = app.load_profile(profile);
if status, error('Failed loading profile.'); end
status = app.set_detection_zone(1, 2);
if status, error('Failed setting detection zone.'); end
xep.delete_file(xep.XTFILE_TYPE_NOISEMAP_FAST, 0);
xep.delete_file(xep.XTFILE_TYPE_NOISEMAP_SLOW, 0);

disp('Initializing another one.');
status = app.set_sensor_mode('run');
if status, error('Failed starting sensor.'); end
disp('waiting for 130s ...');
pause(130);
status = app.set_sensor_mode('stop');
if status, error('Failed stopping sensor.'); end
disp('Done.')
pause(1);

files = xep.find_all_files();
if ~ismember(xep.XTFILE_TYPE_NOISEMAP_FAST, files)
    error('Did not store a fast noisemap');
end
if ~ismember(xep.XTFILE_TYPE_NOISEMAP_SLOW, files)
    error('Did not store a slow noisemap');
end

short_nm_fast = xep.get_file(xep.XTFILE_TYPE_NOISEMAP_FAST, 0);
short_nm_slow = xep.get_file(xep.XTFILE_TYPE_NOISEMAP_SLOW, 0);

xep.delete_file(xep.XTFILE_TYPE_NOISEMAP_FAST, 0);
xep.delete_file(xep.XTFILE_TYPE_NOISEMAP_SLOW, 0);

% Once more, with a restored previous noisemap.

% Upload noisemap to module flash.
xep.set_file(xep.XTFILE_TYPE_NOISEMAP_FAST, 0, nm_fast);
xep.set_file(xep.XTFILE_TYPE_NOISEMAP_SLOW, 0, nm_slow);
pause(1);

disp('Starting profile with the first noisemap.');
status = app.load_profile(profile);
if status, error('Failed loading profile.'); end

% This must match the detection zone the noisemap was made with for the
% restoring to work.
status = app.set_detection_zone(1, 3);
if status, error('Failed setting detection zone.'); end

display('Initialization should now only be around 20 seconds.');
status = app.set_sensor_mode('run');
if status, error('Failed starting sensor.'); end
pause(20);
display('Initialization done.');


display('Getting a list of noisemaps as they evolve over time.');
for ix=[1:5]
    pause(50);
    status = app.store_noisemap();
    if status, error('Failed storing noisemap.'); end
    % Store noisemap takes around a second to complete, pluss a small varying
    % amount of time for the flash writing to catch up.
    pause(10);
    nms_fast{ix} = xep.get_file(xep.XTFILE_TYPE_NOISEMAP_FAST, 0);
    nms_slow{ix} = xep.get_file(xep.XTFILE_TYPE_NOISEMAP_SLOW, 0);
end

status = app.set_sensor_mode('stop');
if status, error('Failed stopping sensor.'); end

File: utils/XeThru_utils/xeX4Thru_software/ModuleConnector/ModuleConnector-unix-1/python27-x86_64-linux-gnu/pymoduleconnector/examples/X4M200_X4M300_manipulate_noisemap.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
""" \example X4M200_X4M300_manipulate_noisemap.py

#Target module: X4M200,X4M300

#Introduction:  This is an example of how to use noisemap on X4M200 and X4M300 modules.
			   
#Command to run:  "python X4M200_X4M300_manipulate_noisemap.py -d com8 -i x4m300" or "python3 X4M200_X4M300_manipulate_noisemap.py -d com8 i x4m300"
                 Using TCP server address as device name is also supported, e.g. 
                 "python X4M200_sleep_record.py -d tcp://192.168.1.169:3000".

"""

from __future__ import print_function

import time
from optparse import OptionParser
import sys

import pymoduleconnector
from pymoduleconnector.ids import *


parser = OptionParser()
parser.add_option(
    "-d",
    "--device",
    dest="device_name",
    help="device file to use, example: python %s -d COM4" % sys.argv[0],
    metavar="FILE")

parser.add_option(
    "-i",
    "--interface",
    dest="interface",
    default="x4m300",
    help="Interface to use. x4m300 or x4m200",
    metavar="IF")


(options, args) = parser.parse_args()

mc = pymoduleconnector.ModuleConnector(options.device_name, 0)
# Get the XEP interface
xep = mc.get_xep()

def ls():
    files = xep.find_all_files()
    return set(zip(files.file_type_items, files.file_identifier_items))

# Get the app interface and set profile
if options.interface == 'x4m300':
    app = mc.get_x4m300()
    profile = XTS_ID_APP_PRESENCE_2
elif options.interface == 'x4m200':
    app = mc.get_x4m200()
    profile = XTS_ID_APP_RESPIRATION_2
else:
    raise SystemExit("Interface not recognized.")

# Load a profile. Noisemap controls are stored associated with this profile,
# meaning setting/getting it without loading a profile doesn't work.
app.load_profile(profile)

# Set a detection zone. The noisemap requires a matching detection zone.
app.set_detection_zone(1.0, 3.0)

# Set noisemap controls
ctrl = app.get_noisemap_control()
# Enable use of noisemap
ctrl |=  XTID_NOISEMAP_CONTROL_ENABLE
# Enable adaptive noisemap
ctrl |=  XTID_NOISEMAP_CONTROL_ADAPTIVE
# Don't reinitialize noisemap if stored, and store it after
# initialization if not present.
ctrl &= ~XTID_NOISEMAP_CONTROL_INIT_ON_RESET
app.set_noisemap_control(ctrl)

files = ls()

if (XTFILE_TYPE_NOISEMAP_FAST, 0) in files:
    xep.delete_file(XTFILE_TYPE_NOISEMAP_FAST, 0)
if (XTFILE_TYPE_NOISEMAP_SLOW, 0) in files:
    xep.delete_file(XTFILE_TYPE_NOISEMAP_SLOW, 0)

print('Initializing a noisemap. Vacate the premises.Please wait for ~2min.')
time.sleep(5)
status = app.set_sensor_mode(XTID_SM_RUN, 0)
time.sleep(130)
status = app.set_sensor_mode(XTID_SM_STOP, 0)
print('Done.')
time.sleep(1)

files = ls()

if (XTFILE_TYPE_NOISEMAP_FAST, 0) not in files:
    raise SystemExit("Did not store a fast noisemap")
if (XTFILE_TYPE_NOISEMAP_SLOW, 0) not in files:
    raise SystemExit("Did not store a slow noisemap")

nm_fast = xep.get_file(XTFILE_TYPE_NOISEMAP_FAST, 0)
nm_slow = xep.get_file(XTFILE_TYPE_NOISEMAP_SLOW, 0)

status = app.load_profile(profile)
status = app.set_detection_zone(1, 2)
xep.delete_file(XTFILE_TYPE_NOISEMAP_FAST, 0)
xep.delete_file(XTFILE_TYPE_NOISEMAP_SLOW, 0)

print('Initializing another one.Please wait for 130s')
app.set_sensor_mode(XTID_SM_RUN, 0)
time.sleep(130)
app.set_sensor_mode(XTID_SM_STOP, 0)
print('Done.')
time.sleep(1)

files = ls()

if (XTFILE_TYPE_NOISEMAP_FAST, 0) not in files:
    raise SystemExit("Did not store a fast noisemap")
if (XTFILE_TYPE_NOISEMAP_SLOW, 0) not in files:
    raise SystemExit("Did not store a slow noisemap")

short_nm_fast = xep.get_file(XTFILE_TYPE_NOISEMAP_FAST, 0)
short_nm_slow = xep.get_file(XTFILE_TYPE_NOISEMAP_SLOW, 0)

xep.delete_file(XTFILE_TYPE_NOISEMAP_FAST, 0)
xep.delete_file(XTFILE_TYPE_NOISEMAP_SLOW, 0)

# Once more, with a restored previous noisemap.

# Upload noisemap to module flash.
xep.set_file(XTFILE_TYPE_NOISEMAP_FAST, 0, nm_fast)
xep.set_file(XTFILE_TYPE_NOISEMAP_SLOW, 0, nm_slow)
time.sleep(1)

print('Starting profile with the first noisemap.')
app.load_profile(profile)

# This must match the detection zone the noisemap was made with for the
# restoring to work.
app.set_detection_zone(1.0, 3.0)

print('Initialization should now only be around 20 seconds.')
app.set_sensor_mode(XTID_SM_RUN, 0)
time.sleep(20)
print('Initialization done.')

print('Getting a list of noisemaps as they evolve over time.')

nms_fast = []
nms_slow = []
for ix in range(5):
    time.sleep(50)
    app.store_noisemap()
    # Store noisemap takes around a second to complete, pluss a varying
    # amount of time for the flash writing to catch up.
    time.sleep(10)
    nms_fast.append(xep.get_file(XTFILE_TYPE_NOISEMAP_FAST, 0))
    nms_slow.append(xep.get_file(XTFILE_TYPE_NOISEMAP_SLOW, 0))

app.set_sensor_mode(XTID_SM_STOP, 0)
